### PR TITLE
Make convertSemesters as pure function

### PR DESCRIPTION
**Pull request by SamChou19815**
### Summary 

The `firestoreCourseToAppCourse` is still doing too many things. After talking with @hanywang2, looks like the predecessor of this function `convertCourse` is doing two things at once: converting data from firestore and converting data from course roster. This is very bad because it makes it very hard to reason about what kind of data we are dealing with, and it caused a tight coupling between data conversion and component logic.

This PR starts to unentangle it. It's obvious from context that `convertSemesters` is only dealing with firebase data, so we can see how we can simplify it. We know in the firestore data that every course has a unique id and a color, so there is no need to call some code to create color and id for it. I believe those code is written for course roster data.

To avoid doing too many things at once in this PR, I still keep the ugly callback structure in `firestoreCourseToAppCourse` (the `incrementID` and `addColor` callback), but make the callback passed into `firestoreCourseToAppCourse` from `convertSemesters` no longer having side effect. After this, `convertCourse` will finally become a pure function!

### Test Plan 

I added some assertions that this is safe. You can load and look at the console that none of the error newly added is thrown.

### CODE REVIEW

File: src/components/Dashboard.vue [hunk 1]
```diff
@@ -243,19 +249,28 @@ export default Vue.extend({
       this.isOpeningRequirements = !this.isOpeningRequirements;
     },
 
-    convertSemesters(firebaseSems: readonly FirestoreSemester[]) {
-      const semesters = firebaseSems.map(firebaseSem => {
-        const appSemester = firestoreSemesterToAppSemester(
+    convertSemesters(
+      firebaseSems: readonly FirestoreSemester[],
+      subjectColors: { readonly [subject: string]: string },
```
Comment (hanywang2) [inline src/components/Dashboard.vue:254]: What's the purpose of the refactor where subjectColors is applied to convertSemesters? How i this better than the former solution

**Review by hanywang2 — COMMENTED** [review]
Question

File: src/components/Dashboard.vue (hunk 1 above)
Comment (SamChou19815) [reply src/components/Dashboard.vue:254]: Then convertSemesters is no longer entangled with `addColor`, which has some side effect (doing some firestore operations) that's hard to reason about.

Comment (SamChou19815) [reply src/components/Dashboard.vue:254]: @hanywang2

**Comment by SamChou19815** [comment]
subsumed in #185.

**Pull request closed without merging**